File: resilient/resilient-annotation-processor/src/main/java/ru/tinkoff/kora/resilient/annotation/processor/aop/FallbackKoraAspect.java
package ru.tinkoff.kora.resilient.annotation.processor.aop;

import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.CodeBlock;
import ru.tinkoff.kora.annotation.processor.common.MethodUtils;
import ru.tinkoff.kora.aop.annotation.processor.KoraAspect;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.ExecutableElement;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

import static com.palantir.javapoet.CodeBlock.joining;

public class FallbackKoraAspect implements KoraAspect {

    private static final ClassName ANNOTATION_TYPE = ClassName.get("ru.tinkoff.kora.resilient.fallback.annotation", "Fallback");

    private final ProcessingEnvironment env;

    public FallbackKoraAspect(ProcessingEnvironment env) {
        this.env = env;
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return Set.of(ANNOTATION_TYPE.canonicalName());
    }

    @Override
    public Set<ClassName> getSupportedAnnotationClassNames() {
        return Set.of(ANNOTATION_TYPE);
    }

    @Override
    public ApplyResult apply(ExecutableElement method, String superCall, AspectContext aspectContext) {
        final Optional<? extends AnnotationMirror> mirror = method.getAnnotationMirrors().stream().filter(a -> a.getAnnotationType().toString().equals(ANNOTATION_TYPE.canonicalName())).findFirst();
        final FallbackMeta fallback = mirror.flatMap(a -> a.getElementValues().entrySet().stream()
                .filter(e -> e.getKey().getSimpleName().contentEquals("method"))
                .map(e -> String.valueOf(e.getValue().getValue())).findFirst()
                .filter(v -> !v.isBlank()))
            .map(v -> FallbackMeta.ofFallbackMethod(v, method))
            .orElseThrow(() -> new IllegalStateException("Method argument for @Fallback is mandatory!"));
        final String name = mirror.flatMap(a -> a.getElementValues().entrySet().stream()
                .filter(e -> e.getKey().getSimpleName().contentEquals("value"))
                .map(e -> String.valueOf(e.getValue().getValue())).findFirst())
            .orElseThrow();

        var managerType = env.getTypeUtils().getDeclaredType(env.getElementUtils().getTypeElement("ru.tinkoff.kora.resilient.fallback.FallbackManager"));
        var fieldManager = aspectContext.fieldFactory().constructorParam(managerType, List.of());
        var fallbackType = env.getTypeUtils().getDeclaredType(env.getElementUtils().getTypeElement("ru.tinkoff.kora.resilient.fallback.Fallback"));
        var fieldFallback = aspectContext.fieldFactory().constructorInitialized(
            fallbackType, CodeBlock.of("$L.get($S)", fieldManager, name));

        final CodeBlock body;
        if (MethodUtils.isMono(method)) {
            body = buildBodyMono(method, fallback, superCall, fieldFallback);
        } else if (MethodUtils.isFlux(method)) {
            body = buildBodyFlux(method, fallback, superCall, fieldFallback);
        } else if (MethodUtils.isFuture(method)) {
            body = buildBodyFuture(method, fallback, superCall, fieldFallback);
        } else {
            body = buildBodySync(method, fallback, superCall, fieldFallback);
        }

        return new ApplyResult.MethodBody(body);
    }

    private CodeBlock buildBodySync(ExecutableElement method, FallbackMeta fallbackCall, String superCall, String fieldFallback) {
        final String fallbackMethod = fallbackCall.call();
        if (MethodUtils.isVoid(method)) {
            final CodeBlock superMethod = buildMethodCall(method, superCall);
            return CodeBlock.builder().add("""
                try {
                    $L;
                } catch (Throwable _e) {
                    if ($L.canFallback(_e)) {
                        $L;
                    } else {
                        throw _e;
                    }
                }
                """, superMethod.toString(), fieldFallback, fallbackMethod).build();
        } else {
            final CodeBlock superMethod = buildMethodCall(method, superCall);
            return CodeBlock.builder().add("""
                try {
                    return $L;
                } catch (Throwable _e) {
                    if ($L.canFallback(_e)) {
                        return $L;
                    } else {
                        throw _e;
                    }
                }
                """, superMethod.toString(), fieldFallback, fallbackMethod).build();
        }
    }

    private CodeBlock buildBodyFuture(ExecutableElement method, FallbackMeta fallbackCall, String superCall, String fieldFallback) {
        final String fallbackMethod = fallbackCall.call();
        final CodeBlock superMethod = buildMethodCall(method, superCall);

        return CodeBlock.builder().add("""
                return $L.exceptionallyCompose(_e -> {
                    var _cause = _e;
                    if (_cause instanceof $T ce) {
                        _cause = ce.getCause();
                    }
                    if ($L.canFallback(_cause)) {
                        return $L;
                    }
                    return $T.failedFuture(_cause);
                });""", superMethod.toString(), CompletionException.class, fieldFallback, fallbackMethod,
            CompletableFuture.class).build();
    }

    private CodeBlock buildBodyMono(ExecutableElement method, FallbackMeta fallbackCall, String superCall, String fieldFallback) {
        final CodeBlock superMethod = buildMethodCall(method, superCall);
        final String fallbackMethod = fallbackCall.call();
        return CodeBlock.builder().add("""
            return $L.onErrorResume($L::canFallback, _e -> $L);
            """, superMethod.toString(), fieldFallback, fallbackMethod).build();
    }

    private CodeBlock buildBodyFlux(ExecutableElement method, FallbackMeta fallbackCall, String superCall, String fieldFallback) {
        final CodeBlock superMethod = buildMethodCall(method, superCall);
        final String fallbackMethod = fallbackCall.call();
        return CodeBlock.builder().add("""
            return $L.onErrorResume($L::canFallback, _e -> $L);
            """, superMethod.toString(), fieldFallback, fallbackMethod).build();
    }

    private CodeBlock buildMethodCall(ExecutableElement method, String call) {
        return method.getParameters().stream().map(p -> CodeBlock.of("$L", p)).collect(joining(", ", call + "(", ")"));
    }
}

File: resilient/resilient-symbol-processor/src/main/kotlin/ru/tinkoff/kora/resilient/symbol/processor/aop/FallbackKoraAspect.kt
package ru.tinkoff.kora.resilient.symbol.processor.aop

import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.MemberName
import ru.tinkoff.kora.aop.symbol.processor.KoraAspect
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotations
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValue
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFlow
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFlux
import ru.tinkoff.kora.ksp.common.FunctionUtils.isCompletionStage
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFuture
import ru.tinkoff.kora.ksp.common.FunctionUtils.isMono
import ru.tinkoff.kora.ksp.common.FunctionUtils.isSuspend
import ru.tinkoff.kora.ksp.common.FunctionUtils.isVoid
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import java.util.concurrent.CompletionStage
import java.util.concurrent.Future

class FallbackKoraAspect(val resolver: Resolver) : KoraAspect {

    companion object {
        private val ANNOTATION_TYPE = ClassName("ru.tinkoff.kora.resilient.fallback.annotation", "Fallback")
    }

    override fun getSupportedAnnotationTypes(): Set<String> {
        return setOf(ANNOTATION_TYPE.canonicalName)
    }

    override fun apply(ksFunction: KSFunctionDeclaration, superCall: String, aspectContext: KoraAspect.AspectContext): KoraAspect.ApplyResult {
        if (ksFunction.isFuture()) {
            throw ProcessingErrorException("@Fallback can't be applied for types assignable from ${Future::class.java}", ksFunction)
        } else if (ksFunction.isCompletionStage()) {
            throw ProcessingErrorException("@Fallback can't be applied for types assignable from ${CompletionStage::class.java}", ksFunction)
        } else if (ksFunction.isMono()) {
            throw ProcessingErrorException("@Fallback can't be applied for types assignable from ${CommonClassNames.mono}", ksFunction)
        } else if (ksFunction.isFlux()) {
            throw ProcessingErrorException("@Fallback can't be applied for types assignable from ${CommonClassNames.flux}", ksFunction)
        }

        val annotation = ksFunction.findAnnotations(ANNOTATION_TYPE).first()

        val fallbackName = annotation.findValue<String>("value")!!
        val fallback = annotation.asFallback(ksFunction)

        val managerType = resolver.getClassDeclarationByName("ru.tinkoff.kora.resilient.fallback.FallbackManager")!!.asType(listOf())
        val fieldManager = aspectContext.fieldFactory.constructorParam(managerType, listOf())
        val fallbackType = resolver.getClassDeclarationByName("ru.tinkoff.kora.resilient.fallback.Fallback")!!.asType(listOf())
        val fieldFallback = aspectContext.fieldFactory.constructorInitialized(
            fallbackType,
            CodeBlock.of("%L[%S]", fieldManager, fallbackName)
        )

        val body = if (ksFunction.isFlow()) {
            buildBodyFlow(ksFunction, fallback, superCall, fieldFallback)
        } else if (ksFunction.isSuspend()) {
            buildBodySync(ksFunction, fallback, superCall, fieldFallback)
        } else {
            buildBodySync(ksFunction, fallback, superCall, fieldFallback)
        }

        return KoraAspect.ApplyResult.MethodBody(body)
    }

    private fun buildBodySync(
        method: KSFunctionDeclaration, fallbackCall: FallbackMeta, superCall: String, fieldFallback: String
    ): CodeBlock {
        val prefix = if (method.isVoid()) "" else "return "
        val superMethod = buildMethodCall(method, superCall)
        return CodeBlock.builder().add(
            """
            ${prefix}try {
                %L
            } catch (_e: Throwable) {
                if(%L.canFallback(_e)) {
                    %L
                } else {
                    throw _e
                }
            }
            """.trimIndent(), superMethod.toString(), fieldFallback, fallbackCall.call()
        ).build()
    }

    private fun buildBodyFlow(
        method: KSFunctionDeclaration, fallbackCall: FallbackMeta, superCall: String, fieldFallback: String
    ): CodeBlock {
        val flowMember = MemberName("kotlinx.coroutines.flow", "flow")
        val catchMember = MemberName("kotlinx.coroutines.flow", "catch")
        val emitMember = MemberName("kotlinx.coroutines.flow", "emitAll")
        val superMethod = buildMethodCall(method, superCall)
        return CodeBlock.builder().add(
            """
            return %M {
                %M(%L)
            }.%M { _e ->
                if (%L.canFallback(_e)) {
                    %M(%L)
                } else {
                    throw _e
                }
            }
            """.trimIndent(), flowMember, emitMember, superMethod.toString(), catchMember, fieldFallback, emitMember, fallbackCall.call()
        ).build()
    }

    private fun buildMethodCall(method: KSFunctionDeclaration, call: String): CodeBlock {
        return CodeBlock.of(method.parameters.asSequence().map { p -> CodeBlock.of("%L", p) }.joinToString(", ", "$call(", ")"))
    }

    private fun buildMethodCallable(method: KSFunctionDeclaration, call: String): CodeBlock {
        val callableMember = MemberName("java.util.concurrent", "Callable")
        return CodeBlock.builder().add("%M { %L }", callableMember, buildMethodCall(method, call)).build()
    }
}
